core/src/main/java/com/linecorp/armeria/client/retry/RetryRuleBuilder.java

```diff
@@ -0,0 +1,282 @@
+        return onException(UnprocessedRequestException.class);
+    }
+
+    /**
+     * Sets the {@linkplain Backoff#ofDefault() default backoff} and returns a newly created {@link RetryRule}.
+     */
+    public RetryRule thenBackoff() {
+        return thenBackoff(Backoff.ofDefault());
+    }
+
+    /**
+     * Sets the specified {@link Backoff} and returns a newly created {@link RetryRule}.
+     */
+    public RetryRule thenBackoff(Backoff backoff) {
+        requireNonNull(backoff, "backoff");
+        return build(RetryRuleDecision.retry(backoff));
+    }
+
+    /**
+     * Returns a newly created {@link RetryRule} that never retries.
+     */
+    public RetryRule thenNoRetry() {
+        return build(RetryRuleDecision.noRetry());
+    }
+
+    private RetryRule build(RetryRuleDecision decision) {
+        final Set<HttpMethod> methods = Sets.immutableEnumSet(this.methods);
+        final Set<HttpStatusClass> statusClasses = Sets.immutableEnumSet(statusClassesBuilder.build());
+        final Set<HttpStatus> statuses = statusesBuilder.build();
+        final Predicate<HttpStatus> statusFilter = this.statusFilter;
+        final Predicate<Throwable> exceptionFilter = this.exceptionFilter;
+
+        if (decision != RetryRuleDecision.noRetry() && exceptionFilter == null && statusFilter == null &&
+            statuses.isEmpty() && statusClasses.isEmpty()) {
+            throw new IllegalStateException(
+                    "Should set at least one of status, status class and an expected exception type " +
+                    "if a backoff was set.");
+        }
+        final CompletableFuture<RetryRuleDecision> decisionFuture;
+        if (decision == RetryRuleDecision.DEFAULT) {
+           decisionFuture = DEFAULT_DECISION;
+        } else {
+            decisionFuture = CompletableFuture.completedFuture(decision);
+        }
+        return (ctx, cause) -> {
+            if (!methods.contains(ctx.request().method())) {
+                return NEXT;
+            }
+
+            if (cause != null && exceptionFilter != null && exceptionFilter.test(Exceptions.peel(cause))) {
+                return decisionFuture;
+            }
+
+            if (ctx.log().isAvailable(RequestLogProperty.RESPONSE_HEADERS)) {
+                final HttpStatus responseStatus = ctx.log().partial().responseHeaders().status();
+                if (statusClasses != null && statusClasses.contains(responseStatus.codeClass())) {
+                    return decisionFuture;
+                }
+
+                if ((statuses != null && statuses.contains(responseStatus)) ||
+                    (statusFilter != null && statusFilter.test(responseStatus))) {
+                    return decisionFuture;
+                }
+            }
+
+            return NEXT;
+        };
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(this)
+                .add("statusClasses", statusClassesBuilder.build())
```

Comment:
nit: indentation?

Reply:
Fixed. Chaining methods are not aligned vertically in IntelliJ IDEA 2020.1. 😭